src/shared/Core/Trace2.cs

```diff
@@ -164,7 +171,7 @@ internal bool TryGetPipeName(string eventTarget, out string name)
         return false;
     }
 
-    private void TryParseSettings(TextWriter error, IFileSystem fileSystem)
+    private void InitializeWriters()
```

Comment:
**Doesn't need to stop this PR**, but I wonder if in the future we could look to just having `Trace2`'s constructor take in the entire `ICommandContext`?

Then we can use the `IStandardStreams` and `IFileSystem` components here, given the 'no-op until `Initialize` is called' approach to resolving the dependency loop.

```patch
diff --git a/src/shared/Atlassian.Bitbucket.UI.Avalonia/Program.cs b/src/shared/Atlassian.Bitbucket.UI.Avalonia/Program.cs
index e2e231ac..03dcc322 100644
--- a/src/shared/Atlassian.Bitbucket.UI.Avalonia/Program.cs
+++ b/src/shared/Atlassian.Bitbucket.UI.Avalonia/Program.cs
@@ -49,7 +49,7 @@ namespace Atlassian.Bitbucket.UI
             using (var app = new HelperApplication(context))
             {
                 // Initialize TRACE2 system
-                context.Trace2.Initialize(context.Environment, context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
+                context.Trace2.Initialize(context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
 
                 // Write the start and version events
                 context.Trace2.Start(context.ApplicationPath, args);
diff --git a/src/shared/Core/CommandContext.cs b/src/shared/Core/CommandContext.cs
index 723d0f3a..21d72b69 100644
--- a/src/shared/Core/CommandContext.cs
+++ b/src/shared/Core/CommandContext.cs
@@ -92,7 +92,7 @@ namespace GitCredentialManager
 
             Streams = new StandardStreams();
             Trace   = new Trace();
-            Trace2  = new Trace2();
+            Trace2  = new Trace2(this);
 
             if (PlatformUtils.IsWindows())
             {
diff --git a/src/shared/Core/Trace2.cs b/src/shared/Core/Trace2.cs
index 62c75070..ae016e23 100644
--- a/src/shared/Core/Trace2.cs
+++ b/src/shared/Core/Trace2.cs
@@ -37,10 +37,9 @@ public interface ITrace2 : IDisposable
     /// <summary>
     /// Initialize TRACE2 tracing by initializing multi-use fields and setting up any configured target formats.
     /// </summary>
-    /// <param name="environment">The environment</param>
     /// <param name="settings">Enabled format targets and values.</param>
     /// <param name="startTime">Approximate time calling application began executing.</param>
-    void Initialize(IEnvironment environment, Trace2Settings settings, DateTimeOffset startTime);
+    void Initialize(Trace2Settings settings, DateTimeOffset startTime);
 
     /// <summary>
     /// Write Version and Start events.
@@ -67,6 +66,7 @@ public interface ITrace2 : IDisposable
 
 public class Trace2 : DisposableObject, ITrace2
 {
+    private readonly ICommandContext _context;
     private readonly object _writersLock = new object();
     private readonly Encoding _utf8NoBomEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
     private readonly List<ITrace2Writer> _writers = new List<ITrace2Writer>();
@@ -79,7 +79,12 @@ public class Trace2 : DisposableObject, ITrace2
 
     private bool _initialized;
 
-    public void Initialize(IEnvironment environment, Trace2Settings settings, DateTimeOffset startTime)
+    public Trace2(ICommandContext context)
+    {
+        _context = context;
+    }
+
+    public void Initialize(Trace2Settings settings, DateTimeOffset startTime)
     {
         if (_initialized)
         {
@@ -88,7 +93,7 @@ public class Trace2 : DisposableObject, ITrace2
 
         _applicationStartTime = startTime;
         _settings = settings;
-        _sid = SetSid(environment);
+        _sid = SetSid(_context.Environment);
 
         InitializeWriters();
 
@@ -187,20 +192,20 @@ public class Trace2 : DisposableObject, ITrace2
             }
             else if (formatTarget.Value.IsTruthy()) // Write to stderr
             {
-                AddWriter(new Trace2StreamWriter(Console.Error, formatTarget.Key));
+                AddWriter(new Trace2StreamWriter(_context.Streams.Error, formatTarget.Key));
             }
             else if (Path.IsPathRooted(formatTarget.Value)) // Write to file
             {
                 try
                 {
-                    Stream stream = File.Open(formatTarget.Value, FileMode.Append,
+                    Stream stream = _context.FileSystem.OpenFileStream(formatTarget.Value, FileMode.Append,
                         FileAccess.Write, FileShare.ReadWrite);
                     AddWriter(new Trace2StreamWriter(new StreamWriter(stream, _utf8NoBomEncoding,
                         4096, leaveOpen: false), formatTarget.Key));
                 }
                 catch (Exception ex)
                 {
-                    Console.Error.WriteLine($"warning: unable to trace to file '{formatTarget.Value}': {ex.Message}");
+                    _context.Streams.Error.WriteLine($"warning: unable to trace to file '{formatTarget.Value}': {ex.Message}");
                 }
             }
         }
diff --git a/src/shared/Git-Credential-Manager.UI.Avalonia/Program.cs b/src/shared/Git-Credential-Manager.UI.Avalonia/Program.cs
index 31df67ff..ed96319a 100644
--- a/src/shared/Git-Credential-Manager.UI.Avalonia/Program.cs
+++ b/src/shared/Git-Credential-Manager.UI.Avalonia/Program.cs
@@ -47,7 +47,7 @@ namespace GitCredentialManager.UI
             using (var app = new HelperApplication(context))
             {
                 // Initialize TRACE2 system
-                context.Trace2.Initialize(context.Environment, context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
+                context.Trace2.Initialize(context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
 
                 // Write the start and version events
                 context.Trace2.Start(context.ApplicationPath, args);
diff --git a/src/shared/Git-Credential-Manager/Program.cs b/src/shared/Git-Credential-Manager/Program.cs
index 8c4dc554..829f187b 100644
--- a/src/shared/Git-Credential-Manager/Program.cs
+++ b/src/shared/Git-Credential-Manager/Program.cs
@@ -15,7 +15,7 @@ namespace GitCredentialManager
             using (var app = new Application(context))
             {
                 // Initialize TRACE2 system
-                context.Trace2.Initialize(context.Environment, context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
+                context.Trace2.Initialize(context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
 
                 // Write the start and version events
                 context.Trace2.Start(context.ApplicationPath, args);
diff --git a/src/shared/GitHub.UI.Avalonia/Program.cs b/src/shared/GitHub.UI.Avalonia/Program.cs
index 105047c0..04ff038e 100644
--- a/src/shared/GitHub.UI.Avalonia/Program.cs
+++ b/src/shared/GitHub.UI.Avalonia/Program.cs
@@ -49,7 +49,7 @@ namespace GitHub.UI
             using (var app = new HelperApplication(context))
             {
                 // Initialize TRACE2 system
-                context.Trace2.Initialize(context.Environment, context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
+                context.Trace2.Initialize(context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
 
                 // Write the start and version events
                 context.Trace2.Start(context.ApplicationPath, args);
diff --git a/src/shared/GitLab.UI.Avalonia/Program.cs b/src/shared/GitLab.UI.Avalonia/Program.cs
index 3b096ef8..7768fffa 100644
--- a/src/shared/GitLab.UI.Avalonia/Program.cs
+++ b/src/shared/GitLab.UI.Avalonia/Program.cs
@@ -48,7 +48,7 @@ namespace GitLab.UI
             using (var app = new HelperApplication(context))
             {
                 // Initialize TRACE2 system
-                context.Trace2.Initialize(context.Environment, context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
+                context.Trace2.Initialize(context.Settings.GetTrace2Settings(), DateTimeOffset.UtcNow);
 
                 // Write the start and version events
                 context.Trace2.Start(context.ApplicationPath, args);
diff --git a/src/shared/TestInfrastructure/Objects/NullTrace.cs b/src/shared/TestInfrastructure/Objects/NullTrace.cs
index 5be2d497..9b0f995a 100644
--- a/src/shared/TestInfrastructure/Objects/NullTrace.cs
+++ b/src/shared/TestInfrastructure/Objects/NullTrace.cs
@@ -52,8 +52,7 @@ namespace GitCredentialManager.Tests.Objects
     {
         #region ITrace2
 
-        public void Initialize(IEnvironment environment,
-            Trace2Settings settings,
+        public void Initialize(Trace2Settings settings,
             DateTimeOffset startTime) { }
 
         public void Start(string appPath,

```